research/conditional/utils/model_utils.py

```diff
@@ -49,90 +48,96 @@
-                    output.reshape(-1, vocab_size),
-                    gt.reshape(-1).long(),
-                    reduction="none",
-                )
-
-                correct_tokens = gt.long() == output.argmax(dim=-1)
-                correct_tokens = correct_tokens.long().reshape(-1) * mask.reshape(-1)
-                correct_tokens = correct_tokens.sum()
-
-                total_tokens = mask.sum()
-
-            return loss[mask.reshape(-1) == 1], correct_tokens, total_tokens
-
-        return custom_forward
+    num_masked_tokens = mask.sum()
 
     with torch.autocast(
         device_type="cuda", enabled=mixed_precision, dtype=torch.float16
     ):
-        encoder_output = model.encoder(input_tokens)
-        chunged_inputs = torch.chunk(encoder_output, n_chungs, dim=0)
+        encoder_output: torch.Tensor = model.encoder(input_tokens)
+        encoder_output_det = encoder_output.detach()
+        if backward_pass:
+            encoder_output_det.requires_grad = True
+        chunged_inputs = torch.chunk(encoder_output_det, n_chungs, dim=0)
         chunged_non_masked_inputs = torch.chunk(gt_tokens, n_chungs, dim=0)
         chunged_non_masked_masks = torch.chunk(mask, n_chungs, dim=0)
 
-        num_tokens = 0
         total_loss = 0
         total_correct_tokens = 0
-        total_masked_tokens = 0
         for chunged_input, chunged_gt, chunged_mask in zip(
             chunged_inputs, chunged_non_masked_inputs, chunged_non_masked_masks
         ):
-            (
-                partial_loss_output,
-                partial_correct_tokens,
-                partial_masked_tokens,
-            ) = checkpoint(
-                make_custom_forward(), chunged_input, chunged_gt, chunged_mask
-            )
-            num_tokens += partial_loss_output.shape[0]
-            total_loss += partial_loss_output.sum()
+            output = model.head(chunged_input)
+            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
+                chunged_gt = chunged_gt.to(output.device)
+                partial_loss_full = F.cross_entropy(
+                    output.reshape(-1, vocab_size),
+                    chunged_gt.reshape(-1).long(),
+                    reduction="none",
+                )
+                partial_loss = partial_loss_full[chunged_mask.reshape(-1) == 1]
+
+                partial_correct_tokens = chunged_gt.long() == output.argmax(dim=-1)
+                partial_correct_tokens = partial_correct_tokens.long().reshape(
+                    -1
+                ) * chunged_mask.reshape(-1)
+                partial_correct_tokens = partial_correct_tokens.sum()
+
+            if backward_pass:
+                loss = partial_loss.sum() / num_masked_tokens / num_accumulated_batches
+                with torch.autocast(
+                    device_type="cuda", enabled=False, dtype=torch.float16
+                ):
+                    scaler.scale(loss).backward()
```

Comment:
isn't this eqivalent to just deleting this context manager? since it's hardcoded to False

Reply:
This code is wrapped in another context manager that may have it set to true, and here we have to turn it off as per torch standard